src/appleseed/renderer/modeling/aov/normalaov.cpp

```diff
@@ -56,40 +59,59 @@ namespace
     // Normal AOV accumulator.
     //
 
+    const Vector3d zero_normal(0.0, 0.0, 0.0);
+
     class NormalAOVAccumulator
-      : public AOVAccumulator
+      : public UnfilteredAOVAccumulator
     {
       public:
-        explicit NormalAOVAccumulator(const size_t index)
-          : AOVAccumulator(index)
+        explicit NormalAOVAccumulator(Image& image)
+          : UnfilteredAOVAccumulator(image)
         {
         }
 
-        virtual void reset() override
+        virtual void on_tile_begin(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y) override
         {
-            m_normal = Vector3f(0.0f);
-        }
+            UnfilteredAOVAccumulator::on_tile_begin(frame, tile_x, tile_y);
 
-        virtual void write(
-            const ShadingPoint&     shading_point,
-            const Camera&           camera) override
-        {
-            if (shading_point.hit())
-                m_normal = Vector3f(shading_point.get_shading_normal());
+            get_tile().clear(Color4f(0.5f, 0.5f, 0.5f, std::numeric_limits<float>::max()));
         }
 
-        virtual void flush(ShadingResult& result) override
+        virtual void write(
+            const PixelContext&         pixel_context,
+            const ShadingPoint&         shading_point,
+            const ShadingComponents&    shading_components,
+            ShadingResult&              shading_result) override
         {
-            result.m_aovs[m_index] =
-                Color4f(
-                    m_normal[0] * 0.5f + 0.5f,
-                    m_normal[1] * 0.5f + 0.5f,
-                    m_normal[2] * 0.5f + 0.5f,
-                    1.0f);
+            const Vector2i& pi = pixel_context.get_pixel_coords();
+
+            // Ignore samples outside the tile.
+            if (outside_tile(pi))
+                return;
+
+            float* p = reinterpret_cast<float*>(
+                get_tile().pixel(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y));
+
+            const float min_sample_squared_distance = p[3];
+            const float sample_squared_distance =
+                squared_distace_to_pixel_center(pixel_context.get_sample_position());
+
+            if (sample_squared_distance < min_sample_squared_distance)
+            {
+                const Vector3d& normal =
```

Comment:
I'd suggest to get rid of `zero_normal` and simplify the code here:

    if (shading_point_hit())
    {
        const Vector3d& n = shading_point.get_shading_normal();
        p[0] = static_cast<float>(n[0]) * 0.5f + 0.5f;
        p[1] = static_cast<float>(n[1]) * 0.5f + 0.5f;
        p[2] = static_cast<float>(n[2]) * 0.5f + 0.5f;
        p[3] = sample_squared_distance;
    }
    else
    {
        p[0] = 0.5f;
        p[1] = 0.5f;
        p[2] = 0.5f;
        p[3] = sample_squared_distance;
    }